diff --ruN a/stablehlo/stablehlo/tests/transforms/stablehlo_aggressive_folder.mlir b/stablehlo/stablehlo/tests/transforms/stablehlo_aggressive_folder.mlir
--- stablehlo/stablehlo/tests/transforms/stablehlo_aggressive_folder.mlir
+++ stablehlo/stablehlo/tests/transforms/stablehlo_aggressive_folder.mlir
@@ -1469,3 +1469,40 @@
   }
   return %1 : tensor<i64>
 }
+
+////////
+// XorOp
+
+// CHECK-LABEL: @xor_fold_splat_cst
+func.func @xor_fold_splat_cst() -> tensor<3x2xi32> {
+  // CHECK: stablehlo.constant dense<1> : tensor<3x2xi32>
+  // CHECK-NOT: stablehlo.xor
+  %operand_1 = stablehlo.constant dense<1> : tensor<3x2xi32>
+  %operand_2 = stablehlo.constant dense<0> : tensor<3x2xi32>
+  %result = stablehlo.xor %operand_1, %operand_2 : tensor<3x2xi32>
+  return %result : tensor<3x2xi32>
+}
+
+// -----
+
+// CHECK-LABEL: @xor_fold_cst
+func.func @xor_fold_cst() -> tensor<3xi32> {
+  // CHECK: stablehlo.constant dense<[0, 0, 3]> : tensor<3xi32>
+  // CHECK-NOT: stablehlo.xor
+  %operand_1 = stablehlo.constant dense<[1, 2, 1]> : tensor<3xi32>
+  %operand_2 = stablehlo.constant dense<[1, 2, 2]> : tensor<3xi32>
+  %result = stablehlo.xor %operand_1, %operand_2 : tensor<3xi32>
+  return %result : tensor<3xi32>
+}
+
+// -----
+
+// CHECK-LABEL: @xor_fold_boolean_cst
+func.func @xor_fold_boolean_cst() -> tensor<4xi1> {
+  // CHECK: stablehlo.constant dense<[true, false, true, false]> : tensor<4xi1>
+  // CHECK-NOT: stablehlo.xor
+  %operand_1 = stablehlo.constant dense<[true, true, false, false]> : tensor<4xi1>
+  %operand_2 = stablehlo.constant dense<[false, true, true, false]> : tensor<4xi1>
+  %result = stablehlo.xor %operand_1, %operand_2 : tensor<4xi1>
+  return %result : tensor<4xi1>
+}
diff --ruN a/stablehlo/stablehlo/tests/transforms/stablehlo_aggressive_simplification.mlir b/stablehlo/stablehlo/tests/transforms/stablehlo_aggressive_simplification.mlir
--- stablehlo/stablehlo/tests/transforms/stablehlo_aggressive_simplification.mlir
+++ stablehlo/stablehlo/tests/transforms/stablehlo_aggressive_simplification.mlir
@@ -2100,12 +2100,78 @@
 /////////
 // XorOp
 
-// CHECK-LABEL: @xor_cst_on_rhs
-func.func @xor_cst_on_rhs(%arg0: tensor<2xi1>) -> tensor<2xi1> {
+// CHECK-LABEL: @xor_cst_false_on_rhs
+// CHECK-SAME: [[ARG:%.+]]: tensor<2xi1>
+func.func @xor_cst_false_on_rhs(%arg0: tensor<2xi1>) -> tensor<2xi1> {
+  // CHECK-NOT: stablehlo-constant
   %cst = stablehlo.constant dense<false> : tensor<2xi1>
+  // CHECK-NOT: stablehlo-xor
   %0 = stablehlo.xor %cst, %arg0 : tensor<2xi1>
-  // CHECK: stablehlo.xor %arg0, %c : tensor<2xi1>
+  // CHECK: return [[ARG]] : tensor<2xi1>
   return %0 : tensor<2xi1>
+}
+
+// -----
+
+// CHECK-LABEL: @xor_cst_true_on_rhs
+// CHECK-SAME: [[ARG:%.+]]: tensor<2xi1>
+func.func @xor_cst_true_on_rhs(%arg0: tensor<2xi1>) -> tensor<2xi1> {
+  // CHECK: [[CST:%.+]] = stablehlo.constant dense<true> : tensor<2xi1>
+  %cst = stablehlo.constant dense<true> : tensor<2xi1>
+  // CHECK: [[RES:%.+]] = stablehlo.xor [[ARG]], [[CST]] : tensor<2xi1>
+  %0 = stablehlo.xor %cst, %arg0 : tensor<2xi1>
+  // CHECK: return [[RES]]
+  return %0 : tensor<2xi1>
+}
+
+// -----
+
+// CHECK-LABEL: @xor_cst_int_on_rhs
+// CHECK-SAME: [[ARG:%.+]]: tensor<2xi32>
+func.func @xor_cst_int_on_rhs(%arg0: tensor<2xi32>) -> tensor<2xi32> {
+  // CHECK: [[CST:%.+]] = stablehlo.constant dense<[1, 2]> : tensor<2xi32>
+  %cst = stablehlo.constant dense<[1, 2]> : tensor<2xi32>
+  // CHECK: [[RES:%.+]] = stablehlo.xor [[ARG]], [[CST]] : tensor<2xi32>
+  %0 = stablehlo.xor %cst, %arg0 : tensor<2xi32>
+  // CHECK: return [[RES]]
+  return %0 : tensor<2xi32>
+}
+
+// -----
+
+// CHECK-LABEL: @xor_same_lhs_rhs
+// CHECK-SAME: [[ARG:%.+]]: tensor<2xi32>
+func.func @xor_same_lhs_rhs(%arg0: tensor<2xi32>) -> tensor<2xi32> {
+  // CHECK: [[RES:%.+]] = stablehlo.constant dense<0> : tensor<2xi32>
+  %0 = stablehlo.xor %arg0, %arg0 : tensor<2xi32>
+  // CHECK: return [[RES]]
+  return %0 : tensor<2xi32>
+}
+
+// -----
+
+// CHECK-LABEL: @xor_zero_like_lhs
+// CHECK-SAME: [[ARG:%.+]]: tensor<3xi32>
+func.func @xor_zero_like_lhs(%arg0: tensor<3xi32>) -> tensor<3xi32> {
+  // CHECK-NOT: stablehlo.constant
+  %0 = stablehlo.constant dense<0> : tensor<3xi32>
+  // CHECK-NOT: stablehlo.xor
+  %1 = stablehlo.xor %0, %arg0 : tensor<3xi32>
+  // CHECK: return [[ARG]]
+  return %1 : tensor<3xi32>
+}
+
+// -----
+
+// CHECK-LABEL: @xor_zero_like_rhs
+// CHECK-SAME: [[ARG:%.+]]: tensor<3xi32>
+func.func @xor_zero_like_rhs(%arg0: tensor<3xi32>) -> tensor<3xi32> {
+  // CHECK-NOT: stablehlo.constant
+  %0 = stablehlo.constant dense<0> : tensor<3xi32>
+  // CHECK-NOT: stablehlo.xor
+  %1 = stablehlo.xor %arg0, %0 : tensor<3xi32>
+  // CHECK: return [[ARG]]
+  return %1 : tensor<3xi32>
 }
 
 // -----
diff --ruN a/stablehlo/stablehlo/transforms/VhloToVersion.cpp b/stablehlo/stablehlo/transforms/VhloToVersion.cpp
--- stablehlo/stablehlo/transforms/VhloToVersion.cpp
+++ stablehlo/stablehlo/transforms/VhloToVersion.cpp
@@ -278,12 +278,13 @@
   LLVM_DEBUG(llvm::dbgs() << "Legal op types for target. " << op << '\n');
 
   // Validate location
-  if (!isLegalLocation(op->getLoc(), targetVersion)) {
-    LLVM_DEBUG(llvm::dbgs()
-               << "Illegal op location for target. " << op << '\n');
-    // TODO: One integrated into ecosystem return an error in these cases.
-  }
-  LLVM_DEBUG(llvm::dbgs() << "Legal op location for target. " << op << '\n');
+  LLVM_DEBUG(
+      if (!isLegalLocation(op->getLoc(), targetVersion)) {
+        llvm::dbgs() << "Illegal op location for target. " << op << '\n';
+        // TODO: One integrated into ecosystem return an error in these cases.
+      } else {
+        llvm::dbgs() << "Legal op location for target. " << op << '\n';
+      });
 
   return true;
 }
diff --ruN a/stablehlo/stablehlo/transforms/optimization/StablehloAggressiveFolder.cpp b/stablehlo/stablehlo/transforms/optimization/StablehloAggressiveFolder.cpp
--- stablehlo/stablehlo/transforms/optimization/StablehloAggressiveFolder.cpp
+++ stablehlo/stablehlo/transforms/optimization/StablehloAggressiveFolder.cpp
@@ -2112,6 +2112,47 @@
   }
 };
 
+struct FoldXorOpPattern : public ShapeOpRewritePattern<XorOp> {
+  using ShapeOpRewritePattern::ShapeOpRewritePattern;
+
+  LogicalResult matchAndRewrite(XorOp op,
+                                PatternRewriter& rewriter) const override {
+    auto resultType = op.getType();
+    auto resultElementType = resultType.getElementType();
+    FailureOr<TypedAttr> result;
+
+    if (resultElementType.isInteger(/*width=*/1)) {
+      result = foldBinaryOpIntOrFloat(rewriter, op, FoldLogicalXor{});
+    } else if (resultElementType.isInteger()) {
+      result = foldBinaryOpIntOrFloat(rewriter, op, FoldBitwiseXor{});
+    } else {
+      return rewriter.notifyMatchFailure(op, "Expected integral element type.");
+    }
+
+    if (failed(result)) return failure();
+    rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(op,
+                                                             result.value());
+    return success();
+  }
+
+  struct FoldLogicalXor {
+    APInt operator()(APInt lhs, APInt rhs) const {
+      return APInt(lhs.getBitWidth(), !lhs.isZero() ^ !rhs.isZero());
+    }
+    std::optional<APFloat> operator()(APFloat lhs, APFloat rhs) const {
+      return std::nullopt;
+    }
+  };
+
+  struct FoldBitwiseXor {
+    APInt operator()(APInt lhs, APInt rhs) const { return lhs ^ rhs; }
+
+    std::optional<APFloat> operator()(APFloat lhs, APFloat rhs) const {
+      return std::nullopt;
+    }
+  };
+};
+
 struct StablehloAggressiveFolderPass
     : public impl::StablehloAggressiveFolderPassBase<
           StablehloAggressiveFolderPass> {
@@ -2176,6 +2217,7 @@
                 FoldTransposeOpPattern,               //
                 FoldWhileOpIfDeadAndPresumedPure,     //
                 FoldWhileOpPattern,                   //
+                FoldXorOpPattern,                     //
                 InlineCaseOpWithConstantBranchIndex,  //
                 LowerBoolSplatConstantsIntoReduceOpRegion>(context, options,
                                                            benefit);
diff --ruN a/stablehlo/stablehlo/transforms/optimization/StablehloAggressiveSimplificationPatterns.td b/stablehlo/stablehlo/transforms/optimization/StablehloAggressiveSimplificationPatterns.td
--- stablehlo/stablehlo/transforms/optimization/StablehloAggressiveSimplificationPatterns.td
+++ stablehlo/stablehlo/transforms/optimization/StablehloAggressiveSimplificationPatterns.td
@@ -682,10 +682,17 @@
 def XorOp_CanonicalizeConstantToRhs
   : CanonicalizeConstantToRhs<StableHLO_XorOp>;
 
-// To consider: xor(X, X) -> 0
-//
-// It's unclear if this is beneficial on hardware vs. adding another constant.
-//
-// def XorOp_FoldToZero
-//   : Pat<(StableHLO_XorOp AnyStaticShapeTensor:$operand, $operand),
-//         (StableHLO_ConstantLike<"0"> $operand)>;
+// Pattern: xor(X, X) -> 0
+def XorOp_FoldToZero
+  : Pat<(StableHLO_XorOp AnyStaticShapeTensor:$operand, $operand),
+        (StableHLO_ConstantLike<"0"> $operand)>;
+
+// Pattern: xor(X, 0) -> X
+def XorOp_RemoveNoop
+  : Pat<(StableHLO_XorOp:$op $lhs, (ConstantLikeMatcher IntZero:$value)),
+        (replaceWithValue $lhs)>;
+
+
+
+
+

